benchmark/benchmark.php

```diff
@@ -98,41 +100,91 @@ function runValgrindPhpCgiCommand(
     array $args,
     ?string $cwd = null,
     bool $jit = false,
-    int $warmup = 0,
     int $repeat = 1,
 ): array {
     global $phpCgi;
 
-    $profileOut = __DIR__ . "/profiles/callgrind.out.$name";
+    $profileOut = __DIR__ . "/profiles/callgrind.$name";
     if ($jit) {
-        $profileOut .= '.jit';
+        $profileOut .= '-jit';
     }
 
     $process = runCommand([
         'valgrind',
         '--tool=callgrind',
         '--dump-instr=yes',
+        '--collect-jumps=yes',
+        '--cache-sim=yes',
+        '--branch-sim=yes',
         "--callgrind-out-file=$profileOut",
+        '--verbose',
         '--',
         $phpCgi,
-        '-T' . ($warmup ? $warmup . ',' : '') . $repeat,
+        '-T' . $repeat,
         '-d max_execution_time=0',
         '-d opcache.enable=1',
         '-d opcache.jit=' . ($jit ? 'tracing' : 'disable'),
         '-d opcache.jit_buffer_size=128M',
         '-d opcache.validate_timestamps=0',
         ...$args,
     ]);
-    $instructions = extractInstructionsFromValgrindOutput($process->stderr);
-    if ($repeat > 1) {
-        $instructions = gmp_strval(gmp_div_q($instructions, $repeat));
+
+    // collect metrics for startup, each benchmark run and shutdown
+    $metricsArr = [];
+    foreach (['startup' => 1, ...range(2, $repeat + 1), 'shutdown' => ''] as $k => $kCallgrindOut) {
+        $profileOutSpecific = $profileOut . '.' . $k;
+        rename($profileOut . ($kCallgrindOut === '' ? '' : '.' . $kCallgrindOut), $profileOutSpecific);
+
+        $metricsArr[$k] = extractMetricsFromCallgrindFile($profileOutSpecific);
     }
-    return ['instructions' => $instructions];
+
+    // print all collected metrics
+    print_r(array_map(fn ($metrics) => array_map(fn ($v) => number_format($v, 0, '.', '_'), $metrics), $metricsArr));
+
+    // find the fastest benchmark run
+    $bestRunIndex = 0;
+    foreach (range(0, $repeat - 1) as $k) {
+        if ($metricsArr[$k]['Ir'] < $metricsArr[$bestRunIndex]['Ir']) {
+            $bestRunIndex = $k;
+        }
+    }
+
+    // remove non-fastest profiles from artifacts
+    foreach (range(0, $repeat - 1) as $k) {
+        $profileOutSpecific = $profileOut . '.' . $k;
+
+        if ($k !== $bestRunIndex) {
+            unlink($profileOutSpecific);
+        }
+    }
+
+    // annotate profiles for artifacts
+    foreach (['startup', $bestRunIndex, 'shutdown'] as $k) {
+        $profileOutSpecific = $profileOut . '.' . $k;
+
+        runCommand([
+            'callgrind_annotate',
```

Comment:
Any idea how to convince `callgrind_annotate` to resolve all symbols, especially all the onces defined in Sqlite lib? `libsqlite3-dev` is installed.

Reply:
You would need to compile with `CFLAGS="-g -O2"` to get debug symbols. But with multiple repetitions we don't generally care about startup/shutdown. It's not very meaningful for applications ran with opcache. If you want to provide startup/shutdown you can do so without repetitions.

Reply:
![image](https://github.com/php/php-src/assets/2228672/3b0e4e74-3ab3-49e1-8288-61525cd88b4c)

in the example you can see some examples of unresolved functions (from `callgrind.symfony-demo.0` profile). Will `CFLAGS="-g -O2"` really help if other functions are resolved well?

Reply:
Not for SQLite, since we're not compiling its source files.

Reply:
Debug symbols for Debian packages are commonly available via the corresponding `-dbgsym` package.

Reply:
In Ubuntu docker, after apt update, I get only these:
```
# apt-cache search libsqlite
libsqlite3-0 - SQLite 3 shared library
libsqlite3-dev - SQLite 3 development files
libsqlite3-mod-ceph - SQLite3 VFS for Ceph
libsqlite3-mod-ceph-dev - SQLite3 VFS for Ceph (development files)
libsqlite-tcl - SQLite 2 Tcl bindings
libsqlite0 - SQLite 2 shared library
libsqlite0-dev - SQLite 2 development files
libsqlite3-gst - SQLite bindings for GNU Smalltalk
libsqlite3-mod-blobtoxy - SQLite3 extension module for read-only BLOB to X/Y mapping
libsqlite3-mod-csvtable - SQLite3 extension module for read-only access to CSV files
libsqlite3-mod-impexp - SQLite3 extension module for SQL script, XML, JSON and CSV import/export
libsqlite3-mod-rasterlite2 - SQLite 3 module for huge raster coverages
libsqlite3-mod-spatialite - Geospatial extension for SQLite - loadable module
libsqlite3-mod-virtualpg - Loadable dynamic extension to both SQLite and SpatiaLite
libsqlite3-mod-xpath - SQLite3 extension module for querying XML data with XPath
libsqlite3-mod-zipfile - SQLite3 extension module for read-only access to ZIP files
libsqlite3-ocaml - Embeddable SQL Database for OCaml Programs (runtime)
libsqlite3-ocaml-dev - Embeddable SQL Database for OCaml Programs (development)
libsqlite3-tcl - SQLite 3 Tcl bindings
libsqliteodbc - ODBC driver for SQLite embedded database
```

Reply:
Thank you!! That link brought me to https://valgrind.org/docs/manual/manual-core.html#manual-core.debuginfod . Commited in https://github.com/php/php-src/commit/d4fe3ef19a56d9f87409a9a313aa8579c906f062.

Now I still see some missing source for the core linux libs:
```

--------------------------------------------------------------------------------
The following files chosen for auto-annotation could not be found:
--------------------------------------------------------------------------------
  ./csu/../csu/libc-start.c
  ./csu/../sysdeps/nptl/libc_start_call_main.h
  ./csu/./csu/errno-loc.c
  ./ctype/../include/ctype.h
  ./debug/./debug/explicit_bzero_chk.c
  ./elf/../sysdeps/generic/dl-protected.h
  ./elf/../sysdeps/generic/ldsodefs.h
  ./elf/../sysdeps/x86_64/dl-machine.h
  ./elf/../sysdeps/x86_64/dl-runtime.h
  ./elf/../sysdeps/x86_64/dl-trampoline.h
  ./elf/./elf/dl-lookup.c
  ./elf/./elf/dl-runtime.c
  ./io/../sysdeps/posix/getcwd.c
  ./io/../sysdeps/unix/syscall-template.S
...
  ./io/../sysdeps/unix/sysv/linux/write.c
  ./libio/./libio/fileops.c
  ./libio/./libio/iofflush.c
  ./libio/./libio/libioP.h
  ./malloc/./malloc/arena.c
  ./malloc/./malloc/malloc.c
  ./malloc/./malloc/morecore.c
  ./misc/../sysdeps/unix/syscall-template.S
  ./misc/../sysdeps/unix/sysv/linux/brk.c
  ./misc/../sysdeps/unix/sysv/linux/x86_64/syscall.S
  ./misc/./misc/sbrk.c
  ./misc/./misc/syslog.c
...
  ./stdlib/../stdlib/strtol.c
  ./stdlib/../stdlib/strtol_l.c
  ./stdlib/./stdlib/getenv.c
  ./string/../sysdeps/x86_64/multiarch/memchr-avx2.S
...
  /build/sqlite3-sDRO1D/sqlite3-3.37.2/./src/main.c
  /build/sqlite3-sDRO1D/sqlite3-3.37.2/./src/malloc.c
  /build/sqlite3-sDRO1D/sqlite3-3.37.2/./src/mem1.c
  /build/sqlite3-sDRO1D/sqlite3-3.37.2/./src/memjournal.c
  /build/sqlite3-sDRO1D/sqlite3-3.37.2/./src/mutex.c
  /build/sqlite3-sDRO1D/sqlite3-3.37.2/./src/mutex_unix.c
  /build/sqlite3-sDRO1D/sqlite3-3.37.2/./src/notify.c
  /build/sqlite3-sDRO1D/sqlite3-3.37.2/./src/os.c
...
```

any idea?